tpm2_eventlog: add support for PCR0 replay with different StartupLocality

THS-on opened the pull request:
According to the "TCG PC Client Platform Firmware Profile Specification Level 00 Version 1.05 Revision 23" section 10.4.5.3 the startup locality is the starting value of PCR0. This can be currently either 0 or 3.

This implements support for replaying logs where locality is not 0. 

Fixes: https://github.com/tpm2-software/tpm2-tools/issues/3146

JuergenReppSIT commented:
@THS-on pcr0 is set to the locality for every event `EV_NO_ACTION` in `foreach_digest2`. Shouldn't there be a check of the signature (`STARTUP_LOCALITY_SIGNATURE`)?
The locality should be added to the calls of `foreach_digest2` in the unit test test/unit/test_tpm2_eventlog.c

THS-on commented:
> pcr0 is set to the locality for every event `EV_NO_ACTION` in `foreach_digest2`.

Only if the the locality flag it set to > 0, which is the case we need to handle. I know this is somewhat hacky. Ideally I would like to have a unset value, but did not find a good way to do that in C.

> Shouldn't there be a check of the signature (STARTUP_LOCALITY_SIGNATURE)?

Yes the check is now added.

>  The locality should be added to the calls of foreach_digest2 in the unit test test/unit/test_tpm2_eventlog.c

Now fixed.

JuergenReppSIT commented:
@THS-on Some more questions:
* The spec states "In the case where an HCRTM event did not
occur, this event may not be used" for  the StartupLocality event. Doesn't this mean that the PCR0 will only be set to the locality if there is a HCRTM event before?
* You check the first event whether it's NO_ACTION with  StartupLocality and produce an error if the signature is not StartupLocality. Couldn't there be another NO_ACTION event the first event?

THS-on commented:
> The spec states "In the case where an HCRTM event did not
occur, this event may not be used" for the StartupLocality event. Doesn't this mean that the PCR0 will only be set to the locality if there is a HCRTM event before?

If I understand the spec correctly this seems to be the case, is is currently not implemented in the code.

> You check the first event whether it's NO_ACTION with StartupLocality and produce an error if the signature is not StartupLocality. Couldn't there be another NO_ACTION event the first event?

I don't think so, because this is the only other usage for NO_ACTION. Technically there is also TCG_Sp800_155_PlatformId_Event, but that seems to be deprecated. 
But it might be not the first event in the log overall if we can have a HCRTM before. 
Do we have any logs with HCRTM events?

I can move that check into foreach_event2(), which might be a better place and we then just can set a flag if we found a HCRTM event before.

JuergenReppSIT commented:
> Do we have any logs with HCRTM events?
> 
No unfortunately not.
In the tpm2 tool test files we have only EV_S_CRTM_VERSION and EV_NO_ACTION with "Signature: Spec ID Event03" events. 
It would be nice if we could add such a file to the test files.

> I can move that check into foreach_event2(), which might be a better place and we then just can set a flag if we found a HCRTM event before.

yes that would be better. Thanks.

THS-on commented:
>  It would be nice if we could add such a file to the test files.

Do you know if any devices generate actually logs with that event? I have not seen that.

I've now moved the special handling into `foreach_event2(..)`.

JuergenReppSIT commented:
Very good. Now there are no scan-build errors. 
Wouldn't it be better not to create an error. The spec allows vendor specific  EV_NO_ACTION events:  "10.4.5.4 Vendor Specific Events"
So the check could be
```
   if (memcmp(locality_event->Signature, STARTUP_LOCALITY_SIGNATURE, sizeof(STARTUP_LOCALITY_SIGNATURE)) == 0) {
      locality = locality_event->Cases.StartupLocality;
   }
```

"

THS-on commented:
The vendor specific events are in PCR6, but I changed it now not to error.

mayuanchenma commented:
The spec states "In the case where an HCRTM event did not occur, this event may not be used" for the StartupLocality event. 

if (eventhdr->EventType == EV_EFI_HCRTM_EVENT && eventhdr->PCRIndex == 0) {
            found_hcrtm = true;
        }

        /* Handle StartupLocality in replay for PCR0 */
        if (!found_hcrtm && eventhdr->EventType == EV_NO_ACTION && eventhdr->PCRIndex == 0) {
            if (event_size < sizeof(EV_NO_ACTION_STRUCT)) {
                LOG_ERR("EventSize is too small\n");
                return false;
            }

            EV_NO_ACTION_STRUCT *locality_event = (EV_NO_ACTION_STRUCT*)event->Event;

            if (memcmp(locality_event->Signature, STARTUP_LOCALITY_SIGNATURE,
                       sizeof(STARTUP_LOCALITY_SIGNATURE)) == 0) {
                locality = locality_event->Cases.StartupLocality;
            }
        }
The code means that an HCRTM event did occur, this event may not be used? spec is incorrect？@THS-on @JuergenReppSIT

JuergenReppSIT commented:
The comment  "In the case where an HCRTM event did not occur, this event may not be used" 
in the discussion was wrong. The spec states:
"In the case where an HCRTM event occurred, this event may not be used."
So the implementation is correct.

mayuanchenma commented:
Thanks @JuergenReppSIT
